File: src/spectralnorm/spectralnorm.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

static double eval_a(int i, int j) {
  int ij = i + j;
  return 1.0 / (double)((ij * (ij + 1) / 2) + i + 1);
}

static void eval_a_times_u(const double *u, double *au, int n) {
  for (int i = 0; i < n; ++i) {
    double sum = 0.0;
    for (int j = 0; j < n; ++j) {
      sum += eval_a(i, j) * u[j];
    }
    au[i] = sum;
  }
}

static void eval_at_times_u(const double *u, double *au, int n) {
  for (int i = 0; i < n; ++i) {
    double sum = 0.0;
    for (int j = 0; j < n; ++j) {
      sum += eval_a(j, i) * u[j];
    }
    au[i] = sum;
  }
}

static void eval_ata_times_u(const double *u, double *v, double *w, int n) {
  eval_a_times_u(u, w, n);
  eval_at_times_u(w, v, n);
}

int main(int argc, char **argv) {
  int n = argc > 1 ? atoi(argv[1]) : 5500;
  double *u = malloc((size_t)n * sizeof(*u));
  double *v = malloc((size_t)n * sizeof(*v));
  double *w = malloc((size_t)n * sizeof(*w));
  if (u == NULL || v == NULL || w == NULL) {
    free(u);
    free(v);
    free(w);
    return 1;
  }

  for (int i = 0; i < n; ++i) {
    u[i] = 1.0;
    v[i] = 0.0;
    w[i] = 0.0;
  }

  for (int i = 0; i < 10; ++i) {
    eval_ata_times_u(u, v, w, n);
    eval_ata_times_u(v, u, w, n);
  }

  double vbv = 0.0;
  double vv = 0.0;
  for (int i = 0; i < n; ++i) {
    vbv += u[i] * v[i];
    vv += v[i] * v[i];
  }

  printf("%.9f\n", sqrt(vbv / vv));
  free(u);
  free(v);
  free(w);
  return 0;
}

File: src/spectralnorm/spectralnorm.go
package main

import (
	"fmt"
	"math"
	"os"
	"strconv"
)

func evalA(i, j int) float64 {
	ij := i + j
	return 1.0 / float64((ij*(ij+1)/2)+i+1)
}

func evalATimesU(u, au []float64, n int) {
	for i := 0; i < n; i++ {
		sum := 0.0
		for j := 0; j < n; j++ {
			sum += evalA(i, j) * u[j]
		}
		au[i] = sum
	}
}

func evalAtTimesU(u, au []float64, n int) {
	for i := 0; i < n; i++ {
		sum := 0.0
		for j := 0; j < n; j++ {
			sum += evalA(j, i) * u[j]
		}
		au[i] = sum
	}
}

func evalAtATimesU(u, v, w []float64, n int) {
	evalATimesU(u, w, n)
	evalAtTimesU(w, v, n)
}

func main() {
	n := 5500
	if len(os.Args) > 1 {
		if value, err := strconv.Atoi(os.Args[1]); err == nil {
			n = value
		}
	}

	u := make([]float64, n)
	v := make([]float64, n)
	w := make([]float64, n)
	for i := range u {
		u[i] = 1.0
	}

	for i := 0; i < 10; i++ {
		evalAtATimesU(u, v, w, n)
		evalAtATimesU(v, u, w, n)
	}

	vbv := 0.0
	vv := 0.0
	for i := 0; i < n; i++ {
		vbv += u[i] * v[i]
		vv += v[i] * v[i]
	}

	fmt.Printf("%.9f\n", math.Sqrt(vbv/vv))
}
